internal/client/state-db/pruning.go

```diff
@@ -0,0 +1,288 @@
+			windowSize)
+	}
+
+	queue, err := newInMemDeathRowQueue[BlockHash, Key](db, base)
+	if err != nil {
+		return pruningWindow[BlockHash, Key]{}, err
+	}
+
+	return pruningWindow[BlockHash, Key]{
+		queue: queue,
+		base:  base,
+	}, nil
+}
+
+func (rw *pruningWindow[BlockHash, Key]) WindowSize() uint64 {
+	return rw.queue.Len(rw.base)
+}
+
+// Get the hash of the next pruning block
+func (rw *pruningWindow[BlockHash, Key]) NextHash() (*BlockHash, error) {
+	return rw.queue.NextHash()
+}
+
+func (rw *pruningWindow[BlockHash, Key]) isEmpty() bool {
+	return rw.WindowSize() == 0
+}
+
+// Check if a block is in the pruning window and not be pruned yet
+func (rw *pruningWindow[BlockHash, Key]) HaveBlock(hash BlockHash, number uint64) haveBlock {
+	// if the queue is empty or the block number exceed the pruning window, we definitely
+	// do not have this block
+	if rw.isEmpty() || number < rw.base || number >= rw.base+rw.WindowSize() {
+		return haveBlockNo
+	}
+	return rw.queue.HaveBlock(hash, uint(number-rw.base))
+}
+
+// Prune next block. Expects at least one block in the window. Adds changes to `commit`.
+func (rw *pruningWindow[BlockHash, Key]) PruneOne(commit *CommitSet[Key]) error {
+	pruned, err := rw.queue.PopFront(rw.base)
+	if err != nil {
+		return err
+	}
+	if pruned != nil {
+		log.Printf("TRACE: Pruning %v (%v deleted)", pruned.hash, len(pruned.deleted))
+		index := rw.base
+		var deletedKeys []Key
+		for deleted := range pruned.deleted {
+			deletedKeys = append(deletedKeys, deleted)
+		}
+		commit.Data.Deleted = append(commit.Data.Deleted, deletedKeys...)
+		commit.Meta.Inserted = append(commit.Meta.Inserted, HashDBValue[[]byte]{
+			Hash:    toMetaKey(lastPruned, struct{}{}),
+			DBValue: scale.MustMarshal(index),
+		})
+		commit.Meta.Deleted = append(commit.Meta.Deleted, toPruningJournalKey(rw.base))
+		rw.base += 1
+		return nil
+	} else {
+		log.Printf("TRACE: Trying to prune when there's nothing to prune")
+		return ErrBlockUnavailable
+	}
+}
+
+// Add a change set to the window. Creates a journal record and pushes it to `commit`
+func (rw *pruningWindow[BlockHash, Key]) NoteCanonical(hash BlockHash, number uint64, commit *CommitSet[Key]) error {
+	if rw.base == 0 && rw.isEmpty() && number > 0 {
+		// This branch is taken if the node imports the target block of a warp sync.
+		// assume that the block was canonicalized
+		rw.base = number
+		// The parent of the block was the last block that got pruned.
+		commit.Meta.Inserted = append(commit.Meta.Inserted, HashDBValue[[]byte]{
+			Hash:    toMetaKey(lastPruned, struct{}{}),
+			DBValue: scale.MustMarshal(number - 1),
+		})
+	} else if (rw.base + rw.WindowSize()) != number {
+		return ErrInvalidBlockNumber
+	}
+	log.Printf(
+		"TRACE: Adding to pruning window: %v (%v inserted, %v deleted)",
+		hash, len(commit.Data.Inserted), len(commit.Data.Deleted),
+	)
+	var inserted []Key
+	for _, kv := range commit.Data.Inserted {
+		inserted = append(inserted, kv.Hash)
+	}
```

Comment:
Perhaps we should add a comment to remind that these lines only make sense if our `rw.queue.(type) == inMemDeathRowQueue`, what do you think?